main/codesig/src/JvmModel.scala

```diff
@@ -0,0 +1,205 @@
+
+  object InvokeType {
+    case object Static extends InvokeType
+    case object Virtual extends InvokeType
+    case object Special extends InvokeType
+  }
+
+  sealed trait JType {
+    override def toString = pretty
+
+    /**
+     * A pretty Java-esque dot-delimited syntax for serializing JTypes. Much more
+     * readable and familiar than the slash-based JVM bytecode syntax
+     */
+    def pretty: String
+  }
+
+  object JType {
+    implicit val rw: ReadWriter[MethodSig] =
+      stringKeyRW(readwriter[String].bimap(_.toString, _ => ???))
+
+    sealed class Prim(val pretty: String) extends JType
+
+    object Prim extends {
+      def read(s: String) = all(s(0))
+
+      val all: Map[Char, Prim] = Map(
+        'V' -> (V: Prim),
+        'Z' -> (Z: Prim),
+        'B' -> (B: Prim),
+        'C' -> (C: Prim),
+        'S' -> (S: Prim),
+        'I' -> (I: Prim),
+        'F' -> (F: Prim),
+        'J' -> (J: Prim),
+        'D' -> (D: Prim)
+      )
+
+      case object V extends Prim("void")
+      case object Z extends Prim("boolean")
+      case object B extends Prim("byte")
+      case object C extends Prim("char")
+      case object S extends Prim("short")
+      case object I extends Prim("int")
+      case object F extends Prim("float")
+      case object J extends Prim("long")
+      case object D extends Prim("double")
+    }
+
+    case class Arr(val innerType: JType) extends JType {
+      def pretty = innerType.pretty + "[]"
+    }
+
+    object Arr {
+      def read(s: String)(implicit st: SymbolTable) = Arr(JType.read(s.drop(1)))
+    }
+
+    class Cls private[JvmModel] (val name: String) extends JType {
+      assert(!name.contains('/'), s"JType $name contains invalid '/' characters")
+      assert(!name.contains('['), s"JType $name contains invalid '[' characters")
+
+      def pretty = name
+    }
+
+    object Cls {
+      def fromSlashed(s: String)(implicit st: SymbolTable) = st.JCls(s.replace('/', '.'))
+
+      implicit def rw(implicit st: SymbolTable): ReadWriter[Cls] =
+        stringKeyRW(readwriter[String].bimap(_.name, st.JCls(_)))
+
+      implicit val ordering: Ordering[Cls] = Ordering.by(_.name)
+
+      def read(s: String)(implicit st: SymbolTable) = fromSlashed(s)
+    }
+
+    def read(s: String)(implicit st: SymbolTable): JType = s match {
+      case x if Prim.all.contains(x(0)) => Prim.all(x(0))
+      case s if s.startsWith("L") && s.endsWith(";") => Cls.read(s.drop(1).dropRight(1))
+      case s if s.startsWith("[") => Arr.read(s)
+      case s => Cls.read(s)
+    }
+
+    implicit val ordering: Ordering[JType] = Ordering.by(_.pretty)
+  }
+
+  object Desc {
+
+    private val charBitSet = collection.immutable.BitSet.fromSpecific("BCDFIJSZL".map(x => x: Int))
+    def read(s: String)(implicit st: SymbolTable) = {
+      val scala.Array(argString, ret) = s.drop(1).split(')')
+      val args = Array.newBuilder[JType]
+      var index = 0
+      val argStringLength = argString.length
```

Comment:
Since you now the size of the array beforehand, you can use `new Array[JType](argString.length)` and then return `ArraySeq.unsafeWrapArray(args)`